File: sdks/typescript/src/router.ts
// Routing: decides, for each inbound stream target ({host, port}), which upstream
// to dial — a configured route, the fallback upstream, the client-specified target
// (if allowed), or deny.

import { hostMatches, normalizeHost, toArray } from './util.js';

/** A dialable upstream: {host, port} or {unix}. */
export interface UpstreamAddr {
  host?: string;
  port?: number;
  unix?: string;
}

/** Anything that can name an upstream: "host:port", "unix:/path", [host, port], object. */
export type UpstreamInput = string | [string, number | string] | UpstreamAddr | null | undefined;

/** A single configured route. */
export interface RouteConfig {
  host?: string | string[];
  hosts?: string | string[];
  port?: PortPatterns;
  ports?: PortPatterns;
  upstream?: UpstreamInput;
}

/** A single port pattern: number, "80", "*", "1000-2000", [lo, hi], {min, max}. */
export type PortPattern = number | string | [number, number] | { min?: number; max?: number };
/** One pattern or a list of patterns. */
export type PortPatterns = PortPattern | PortPattern[];

/** Client-dial permission block. */
export interface ClientTargetsConfig {
  enabled?: boolean;
  allowHosts?: string | string[];
  allowPorts?: PortPatterns | null;
}

/** Compiled route (internal). */
export interface CompiledRoute {
  order: number;
  hosts: string[];
  ports: PortPatterns | null;
  upstream: UpstreamAddr | null;
}

/** Resolver verdict for one inbound stream target. */
export type RouteDecision =
  | { action: 'route'; route: number; upstream: UpstreamAddr | null }
  | { action: 'dial'; target: UpstreamAddr }
  | { action: 'deny'; reason: string };

function compileUpstream(u: UpstreamInput): UpstreamAddr | null {
  if (!u) return null;
  if (typeof u === 'string') {
    const idx = u.lastIndexOf(':');
    if (u.startsWith('unix:')) return { unix: u.slice(5) };
    return { host: u.slice(0, idx), port: Number(u.slice(idx + 1)) };
  }
  if (Array.isArray(u)) return { host: u[0], port: Number(u[1]) };
  if ((u as UpstreamAddr).unix !== undefined) return { unix: (u as UpstreamAddr).unix };
  return { host: (u as UpstreamAddr).host ?? '127.0.0.1', port: Number((u as UpstreamAddr).port ?? 0) };
}

function portMatches(patterns: PortPatterns, port: number): boolean {
  for (const p of toArray(patterns)) {
    if (typeof p === 'number') {
      if (p === port) return true;
      continue;
    }
    if (Array.isArray(p) || typeof p === 'object') {
      const lo = Array.isArray(p) ? p[0] : p.min ?? 0;
      const hi = Array.isArray(p) ? p[1] : p.max ?? 65535;
      if (port >= Number(lo) && port <= Number(hi)) return true;
      continue;
    }
    const s = String(p);
    if (s === '*' || s === String(port)) return true;
    const dash = s.match(/^(\d+)-(\d+)$/);
    if (dash && port >= Number(dash[1]) && port <= Number(dash[2])) return true;
  }
  return false;
}

/**
 * Compile a routing config into a resolver.
 * @param opts.routes ordered [{host|hosts, port|ports, upstream}]
 * @param opts.fallbackUpstream used when no route matches
 * @param opts.clientTargets {enabled, allowHosts, allowPorts} —
 *   when enabled, otherwise-unrouted streams may dial their requested target
 * @returns {resolve, routes, fallback} — resolve(target) returns
 *   {action:'route', upstream} | {action:'dial', target} | {action:'deny', reason}
 */
export function compileRoutes(
  { routes = [], fallbackUpstream = null, clientTargets = null }: {
    routes?: RouteConfig | RouteConfig[];
    fallbackUpstream?: UpstreamInput;
    clientTargets?: ClientTargetsConfig | boolean | null;
  } = {},
): { resolve: (target: { host?: string | null; port?: number; unix?: string | null }) => RouteDecision; routes: CompiledRoute[]; fallback: UpstreamAddr | null } {
  const compiled: CompiledRoute[] = toArray(routes).map((r, i) => ({
    order: i,
    hosts: toArray(r.host ?? r.hosts ?? []).map((h) => normalizeHost(h)).filter((h): h is string => !!h),
    ports: (r.port ?? r.ports ?? null) as PortPatterns | null,
    upstream: compileUpstream(r.upstream),
  }));

  const fallback = compileUpstream(fallbackUpstream);

  let ct: { allowHosts: string[]; allowPorts: PortPatterns | null } | null = null;
  if (clientTargets) {
    const enabled = clientTargets === true || clientTargets.enabled === true;
    if (enabled) {
      const cfg = clientTargets === true ? {} : clientTargets;
      ct = {
        allowHosts: toArray((cfg as ClientTargetsConfig).allowHosts ?? ['*']),
        allowPorts: (cfg as ClientTargetsConfig).allowPorts ?? null, // null = any
      };
    }
  }

  function resolve(target: { host?: string | null; port?: number; unix?: string | null }): RouteDecision {
    const host = normalizeHost(target?.host);
    const port = Number(target?.port ?? 0);
    for (const r of compiled) {
      const hostOk = r.hosts.length === 0 || (host != null && r.hosts.some((h) => hostMatches(h, host)));
      const portOk = r.ports == null || portMatches(r.ports, port);
      if (hostOk && portOk) {
        return { action: 'route', route: r.order, upstream: r.upstream ?? fallback };
      }
    }
    if (fallback) return { action: 'route', route: -1, upstream: fallback };
    if (ct) {
      const hostOk = host == null || ct.allowHosts.some((p) => hostMatches(p, host));
      const portOk = ct.allowPorts == null || portMatches(ct.allowPorts, port);
      if (hostOk && portOk && (port > 0 || target?.unix)) {
        return { action: 'dial', target: target?.unix ? { unix: target.unix } : { host: target?.host ?? undefined, port } };
      }
      return { action: 'deny', reason: 'client target not allowed' };
    }
    return { action: 'deny', reason: 'no route matched and client targets are disabled' };
  }

  return { resolve, routes: compiled, fallback };
}

File: crates/krymux/src/router.rs
//! Request routing: mapping stream targets to upstreams, with the same
//! semantics as the Node implementation.

use crate::mux::Target;

#[derive(Clone, Debug)]
/// Where a matched stream is forwarded to.
pub struct Upstream {
    pub host: Option<String>,
    pub port: u16,
    pub unix: Option<String>,
}

impl Upstream {
    /// Renders the upstream as a `host:port` or `unix:path` label.
    pub fn label(&self) -> String {
        if let Some(u) = &self.unix {
            format!("unix:{}", u)
        } else {
            format!(
                "{}:{}",
                self.host.as_deref().unwrap_or("127.0.0.1"),
                self.port
            )
        }
    }
}

#[derive(Clone, Debug)]
/// One route: match targets by host and port patterns, forward to an
/// upstream.
pub struct Route {
    pub hosts: Vec<String>,
    pub ports: Vec<PortPattern>,
    pub upstream: Option<Upstream>,
}

#[derive(Clone, Debug)]
/// Port matcher: exact value, inclusive range, or any.
pub enum PortPattern {
    Exact(u16),
    Range(u16, u16),
    Any,
}

#[derive(Clone, Debug, Default)]
/// Policy for dialing client-requested targets directly (no route needed).
pub struct ClientTargets {
    pub enabled: bool,
    pub allow_hosts: Vec<String>,
    pub allow_ports: Vec<PortPattern>,
}

/// Resolver built from compiled routes, a fallback upstream, and the
/// client-target policy.
pub struct Router {
    routes: Vec<Route>,
    fallback: Option<Upstream>,
    client_targets: ClientTargets,
}

/// Normalizes a host for matching: lowercased, brackets and a trailing
/// `:port` stripped, trailing dot removed. Returns `None` when nothing
/// remains.
pub fn normalize_host(h: &str) -> Option<String> {
    let mut s = h.trim().to_lowercase();
    if s.starts_with('[') {
        // [v6]
        let end = s.find(']')?;
        return Some(s[1..end].to_string());
    }
    if let Some(idx) = s.rfind(':') {
        // strip :port when it looks like host:port (no colons in a bare hostname)
        if !s[..idx].is_empty() && s.matches(':').count() == 1 {
            s = s[..idx].to_string();
        }
    }
    let s = s.trim_end_matches('.').to_string();
    if s.is_empty() {
        None
    } else {
        Some(s)
    }
}

/// Matches `host` against a pattern: exact, `*`, or `*.suffix` (one or more
/// labels under the suffix).
pub fn host_matches(pattern: &str, host: &str) -> bool {
    let Some(p) = normalize_host(pattern) else {
        return false;
    };
    let Some(h) = normalize_host(host) else {
        return false;
    };
    if p == "*" || p == h {
        return true;
    }
    if let Some(suffix) = p.strip_prefix("*.") {
        let suf = format!(".{}", suffix);
        return h.ends_with(&suf) && !h[..h.len() - suf.len()].ends_with('.');
    }
    false
}

fn port_matches(patterns: &[PortPattern], port: u16) -> bool {
    patterns.is_empty()
        || patterns.iter().any(|p| match p {
            PortPattern::Any => true,
            PortPattern::Exact(p) => *p == port,
            PortPattern::Range(lo, hi) => port >= *lo && port <= *hi,
        })
}

impl Router {
    /// Builds a router from compiled routes, a fallback upstream, and the
    /// client-target policy.
    pub fn new(
        routes: Vec<Route>,
        fallback: Option<Upstream>,
        client_targets: ClientTargets,
    ) -> Self {
        Router {
            routes,
            fallback,
            client_targets,
        }
    }

    /// Resolves a stream target to a route, a direct dial, or a denial.
    pub fn resolve(&self, target: &Target) -> Decision {
        let host = target.host.as_deref().and_then(normalize_host);
        let port = target.port;
        for (i, r) in self.routes.iter().enumerate() {
            let host_ok = r.hosts.is_empty()
                || host
                    .as_deref()
                    .map(|h| r.hosts.iter().any(|p| host_matches(p, h)))
                    .unwrap_or(false);
            let port_ok = port_matches(&r.ports, port);
            if host_ok && port_ok {
                if let Some(up) = &r.upstream {
                    return Decision::Route {
                        order: i,
                        upstream: up.clone(),
                    };
                }
                if let Some(f) = &self.fallback {
                    return Decision::Route {
                        order: usize::MAX,
                        upstream: f.clone(),
                    };
                }
                return Decision::Deny {
                    reason: "route has no upstream".into(),
                };
            }
        }
        if let Some(f) = &self.fallback {
            return Decision::Route {
                order: usize::MAX,
                upstream: f.clone(),
            };
        }
        if self.client_targets.enabled {
            // unix-socket targets carry no host, so the allow_hosts check above
            // cannot govern them — dialing arbitrary server sockets (e.g.
            // /var/run/docker.sock) must be opt-in, never implied
            if target.unix.is_some() {
                return Decision::Deny {
                    reason: "client-chosen unix sockets are not allowed".into(),
                };
            }
            let host_ok = host
                .as_deref()
                .map(|h| {
                    self.client_targets
                        .allow_hosts
                        .iter()
                        .any(|p| host_matches(p, h))
                })
                .unwrap_or(true);
            let port_ok = port_matches(&self.client_targets.allow_ports, port);
            if host_ok && port_ok && port > 0 {
                return Decision::Dial;
            }
            return Decision::Deny {
                reason: "client target not allowed".into(),
            };
        }
        Decision::Deny {
            reason: "no route matched and client targets are disabled".into(),
        }
    }
}

/// The outcome of resolving a stream target.
pub enum Decision {
    /// Forward to a fixed upstream; `order` is the matching route's position.
    Route { order: usize, upstream: Upstream },
    /// Dial the client-requested target directly.
    Dial,
    /// Refuse the stream, with a reason for the log.
    Deny { reason: String },
}
